library/pkcs7.c

```diff
@@ -0,0 +1,756 @@
+
+    /* Look for certificates, there may or may not be any */
+    mbedtls_x509_crt_init( &signed_data->certs );
+    ret = pkcs7_get_certificates( &p, end_set, &signed_data->certs );
+    if( ret < 0 )
+        return( ret ) ;
+
+    signed_data->no_of_certs = ret;
+
+    /*
+     * Currently CRLs are not supported. If CRL exist, the parsing will fail
+     * at next step of getting signers info and return error as invalid
+     * signer info.
+     */
+
+    signed_data->no_of_crls = 0;
+
+    /* Get signers info */
+    ret = pkcs7_get_signers_info_set( &p, end_set, &signed_data->signers );
+    if( ret < 0 )
+        return( ret );
+
+    signed_data->no_of_signers = ret;
+
+    /* Don't permit trailing data */
+    if ( p != end )
+        ret = MBEDTLS_ERR_PKCS7_INVALID_FORMAT;
+
+    ret = 0;
+    return( ret );
+}
+
+int mbedtls_pkcs7_parse_der( const unsigned char *buf, const int buflen,
+                             mbedtls_pkcs7 *pkcs7 )
+{
+    unsigned char *start;
+    unsigned char *end;
+    size_t len = 0;
+    int ret;
+    int isoidset = 0;
+
+    if( !pkcs7 )
+        return( MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA );
+
+    /* make an internal copy of the buffer for parsing */
+    pkcs7->raw.p = start = mbedtls_calloc( 1, buflen );
+    if( pkcs7->raw.p == NULL )
+    {
+        return( MBEDTLS_ERR_PKCS7_ALLOC_FAILED );
+    }
+    memcpy( start, buf, buflen );
+    pkcs7->raw.len = buflen;
+    end = start + buflen;
+
+    ret = pkcs7_get_content_info_type( &start, end, &pkcs7->content_type_oid );
+    if( ret != 0 )
+    {
+        len = buflen;
+        goto try_data;
+    }
+
+    if( ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_AND_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DIGESTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret =  MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE;
+        goto out;
+    }
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
+        goto out;
+    }
+
+    isoidset = 1;
+    start = start + pkcs7->content_type_oid.len;
+
+    ret = pkcs7_get_next_content_len( &start, end, &len );
+    if( ret != 0 )
+        goto out;
+
+try_data:
+    ret = pkcs7_get_signed_data( start, len, &pkcs7->signed_data );
+    if ( ret != 0 )
+        goto out;
+
+    if ( !isoidset )
+    {
+        pkcs7->content_type_oid.tag = MBEDTLS_ASN1_OID;
+        pkcs7->content_type_oid.len = MBEDTLS_OID_SIZE(MBEDTLS_OID_PKCS7_SIGNED_DATA);
+        pkcs7->content_type_oid.p = (unsigned char *)MBEDTLS_OID_PKCS7_SIGNED_DATA;
```

Comment:
The constant is a string, so it is a pointer. The specific cast to an `unsigned char *` may not be needed but I am leaving it for clarity purposes. Please let me know if you would rather have this allocated memory instead of a hardcoded constant.

Reply:
Casting `MBEDTLS_OID_PKCS7_SIGNED_DATA` to `unsigned char *` is wrong since it's a constant. Its type is `const unsigned char *`. A cast to `unsigned char *` isn't actually invalid C, as far as I recall, but it's misleading since you aren't allowed to modify it.

Please either remove the cast if it isn't necessary, or explain why the cast is necessary.

Reply:
One of the more important uses of `const` vs not having `const` is knowing when a pointer is something that needs to be `free()`d - you can't `free()` a `const char *`. Casting away this `const`-ness loses this information. It looks like `raw.p` would always need to be `free()`d, but `content_type_oid.p` would never need to be `free()`d, so should they actually be the same type? Is one a PKCS#7 buffer and the other a const PKCS#7 buffer? Or even just a PKCS#7 OID?

Reply:
It seems common practice in MbedTLS to have an allocated `mbedtls_asn1_buf` called `raw` and to use other `mbedtls_asn1_buf.p` to point to data in `raw.p`. Specifically, I am looking at x509_crt as an example. When it comes time to free, only the `raw.p` need to be freed since the other `mbedtls_asn1_buf.p` instances contain references to `raw.p`. 

So, I am a bit confused on how the above typecast is much different. I can make a new type for when the `p` field is `const` but then I would be confused as to why we don't use this new type in all the other places where `mbedtls_asn1_buf.p` is a reference to some offset in another allocated buffer like `raw.p` (since modifications to raw should not be allowed either). 
Apologies, I am not trying to start a fuss but I would like to make sure I am not missing something here.